### ISSUE TITLE
Is concurrent async "in" transfers order guaranted?

### ISSUE THREAD

**Issue by sonatique** [libusb/libusb#1077]
Hello, 
I guess the answer is "yes", but I didn't find a clear official answer despite all my seaches. Sorry if I missed it.

Say I submitted N async transfers for the same "in" endpoint, all with infinite timeout and same buffer size (say 65536 bytes). N is typically 16, but my question holds for any N>1.

I use the event handling thread technique (https://libusb.sourceforge.io/api-1.0/group__libusb__asyncio.html#eventthread) running a single loop in a dedicated thread for processing event.

Say that for any reason, this dedicated loop needs a large amount of time in between two consecutive libusb_handle_events calls (could be because the actual callback takes a long time or because there is some CPU cycle starvation at some point).

Say that during this period of time, a lot of data is received, such that several transfer buffers are filled and are now ready to be processed.

What will happen on the next libusb_handle_events call from within the dedicated thread? Is there a guarantee that the first transfer object to be processed will contain the first data receieved, and then that the next one (on the next loop iteration) will contains the next data?

In other words: is there a mechanism that guarantee that transfers are registered in the "event queue" (whatever it is) in the order they are filled and a guarantee that each call to libusb_handle_events will retreive them is that order?

Thanks a lot.

**Comment by mcuee** [comment]
@hjelmn @tormodvolden Sorry I can not answer this one. Please help to answer. Thanks.

**Comment by tormodvolden** [comment]
I don't know either but I guess this is down to details in how the OS is processing the transfers and presenting events to the libusb code in your application. Is this a general question or do you have a specific OS and driver that you are targeting? What kind of transfers? What problem do you want to solve, or avoid?

**Comment by mcuee** [comment]
@sonatique 

Please help to answer the query from Tormod. Thanks.

**Comment by sonatique** [comment]
@tormodvolden Thank you for jumping in. As far as I can understand, this behavior shall not be dependent on the OS implementation details, or it would basically defeat the portability purpose of libusb, in my opinion. Either the order is guaranteed for all implementations (and your application can rely on it) or it is not guarantee (and you cannot rely on it, though it may happen that it is implemented that way by chance on some given OS). This should be made clear, I think.
I am targeting all possible OS / driver with my application. The transfer type is mainly bulk (though the question should apply to all types).
So now: a little bit more background: take the “sam3u_benchmark.c” example code. You’ll see that the current code only submits one transfer at a time, resubmitting it every time it completes. There is this note in a comment, however:
“NOTE: To reach maximum possible performance the program must submit *multiple* transfers here, not just one.
When only one transfer is submitted there is a gap in the bus schedule from when the transfer completes until a new transfer is submitted by the callback. This causes some jitter for isochronous transfers and loss of throughput for bulk transfers. 
This is avoided by queueing multiple transfers in advance, so that the host controller is always kept busy, and will schedule more transfers on the bus while the callback is running for transfers which have completed on the bus.”
So: I tried to follow this advice, though there is not a single tested example of it in libusb repository.
I faced many issues when going from a little proof of concept to a real complex application. One of it was that I observed disorder: some chunk of data were received before others, in a different order than they were sent on the bus. Say I had a first transfer completing with “4 5 6” in the buffer, then another one with “0 1 2 3” in its buffer, when “0 1 2 3 4 5 6” has been actually sent on the bus.
This was unexpected to me, so I asked this question here. 
Later I found out that the issue was occurring because I was simultaneously i) having several submitted transfers with one properly handled single “event loop thread” to process them and ii) using libusb built-in synchronous Control Transfer functions.
Libusb built-in synchronous Control Transfer function are actually implemented by submitting one async transfer and waiting for completion from within the caller thread. So: during some period of time, I had 2 concurrent event handler threads: the caller thread of the Control Transfer would actually process some completed “data” transfer I submitted, and my event handler thread would process the control transfer completion. Of course, when you have more than one thread processing completions of several submitted transfers, the overall order cannot be guaranteed.
As a result, I rewrote all Control Transfer functions I needed to manually submit transfers in a way that only my event thread would process their completion. Therefore, at every instant I only have one single event thread. Since then, I did not observe disorder anymore.
Rewriting everything was some heavy work and I felt I really was in an undocumented territory: on one hand, some comment in an example program encouraged me to use multiple transfer submissions, but on the other hand, doing it for real “naively” triggers a lot of issue for which there is no help / documentation / example.
So: after I found out that the disorder I observed was caused by multiple event threads acting concurrently, I decided to leave my question open here since I still think that the behavior shall be better documented, and that what is guarantee and what is not shall be better described.
In conclusion: as of today, I managed by myself to not have issues anymore, but given the benefit of using multiple transfers, I think libusb should go beyond a simple comment burden deep in an example code and really provide full working example and improved documentation related to this.
I hope this helps, thanks again.

**Comment by tormodvolden** [comment]
Thanks for the good explanation and context.

How I have seen this is that your application is not guaranteed to receive the transfer completions in order, but the transfer requests will always be sent to the device in the order you submit them. When you receive the transfer completion, it is accompanied by the user_data pointer that you provided in the request so that you can identify and match it to your submitted requests and thus e.g. copy the received data into the right buffer location of your application.

I can imagine how mixing synchronous and synchronous functions ends up with several event handlers although I haven't looked at it. Maybe we should warn about this. So far our sync functions documentation says "Another issue is that by tying up the thread with that single transaction there is no possibility of performing I/O with multiple endpoints and/or multiple devices simultaneously, unless you resort to creating one thread per transaction."

Other relevant documentation is https://libusb.sourceforge.io/api-1.0/group__libusb__asyncio.html#asyncevent which explains how to use an event handling thread, but it doesn't mention possible interference with sync functions. It communicates there should be only one event handling thread though.

Anyway, if we say the completion event order is not guaranteed, and you have to rely on user_data, the issue with multiple event threads that you experienced is not really a problem.

For the example application, please submit one, that would be nice to have. We have been discussing isochronous transfer benchmark tools where this would be needed, I guess related to the sam3u example (#962 ?).

**Comment by tormodvolden** [comment]
The user_data pointer that you provided to e.g. libusb_fill_bulk_stream_transfer() will be in the libusb_transfer struct that is handed to your callback on completion.

**Comment by tormodvolden** [comment]
> As far as I understand, the order in which you submit the transfer doesn’t really matter: there is no “First submitted, first to be used” policy anywhere, so I could as well have used numbering 3, 2, 4, 1,

Are you saying that if you submit them in order 1,2,3,4, you cannot be sure the device will receive them in that order? I thought that was guaranteed but I could be wrong. Anyway that shouldn't be a problem, unless your device needs to receive and process the requests in the defined order. Is this the case?

**Comment by tormodvolden** [comment]
> Since I cannot rely on user data to order USB fragments, I have to rely on the order in which I receive them to order small buffers into the larger buffer.

Yes, you can rely on user_data to order the completed transfers. This is exactly how you can use it. If you submitted requests with user_data 1,2,3,4 you can now read out the user_data from the completed transfers and copy the data buffer into large_buffer[i] where _i_ is derived from user_data.

**Comment by sonatique** [comment]
> > As far as I understand, the order in which you submit the transfer doesn’t really matter: there is no “First submitted, first to be used” policy anywhere, so I could as well have used numbering 3, 2, 4, 1,
> 
> Are you saying that if you submit them in order 1,2,3,4, you cannot be sure the device will receive them in that order? I thought that was guaranteed but I could be wrong. Anyway that shouldn't be a problem, unless your device needs to receive and process the requests in the defined order. Is this the case?

We are talking about "in" transfer here exclusively. I don't think the concept of "the device will receive them" makes sense. I understand that when submitting "in" transfer, all that happen is that there are some structure that are registered at the low level OS system in order to receive incomming data. Nothing is really sent to the device (appart from USB poll). If you plug a USB sniffer you won't see anything when submitting in transfers.

For out transfers however, yes, I guess the order is guranteed.

**Comment by martinling** [comment]
There is plenty of code out there which assumes that multiple transfers submitted _on the same endpoint_ will be completed in order. Two examples that I'm familiar with which do this are libhackrf, and the fx2lafw driver in sigrok. It would be very surprising to find that this code is incorrect. Certainly to the best of my knowledge this approach has never resulted in out-or-order data being seen.

In both those examples, the transfers all use the same `user_data` pointer, with no per-transfer differences, and the order of completion is assumed to be the order of submission.

I wouldn't expect completion ordering to match submission ordering for transfers on different endpoints, but it would be quite a surprise for libusb to announce that there is no ordering guarantee at all.

**Comment by sonatique** [comment]
@tormodvolden @martinling : thank you for your explanations and interest in this subject.
@martinling : in my case, the fact that the first submitted in transfer is the first to be callback-ed for when data arrives on the bus is not important (it is a good thing to know that this is guaranteed anyway). 
My concern is about whether the first transfer that have its callback called will contains the first bytes actually transferred on the bus, for a given endpoint and assuming there is a single thread doing the event loop.
As I understand from @martinling "algorithm" this should be true. 
I can see that since order of completion is guaranteed to be the order of submission, altogether we have: first submitted transfer contains the first part of the transmitted data when it completes. If we’re sure that there is only one thread processing all events, then this is sufficient, we don’t need to use User Data to re-order things.
However, if at some point we have more than one thread processing the even loop (such as having one main loop + a certain number of “sync calls” that introduces new thread into the game, then order could be jammed somehow, but using UserData to track order of submission would help sorting things out.
I hope I got everything right, thanks again!

**Comment by martinling** [comment]
Hi @sonatique,

Something you may not be aware of is that there is an event handling lock, which can only be held by one thread at a time. When you call `libusb_handle_events` it takes this lock, and releases it when it returns. So even if you have multiple threads handling events (e.g. jumping in to make synchronous calls), the overall ordering of completions should be the same.

That's a very rough explanation only - this is all explained in a lot more detail here:
https://libusb.sourceforge.io/api-1.0/libusb_mtasync.html

**Comment by sonatique** [comment]
> Something you may not be aware of is that there is an event handling lock, which can only be held by one thread at a time. When you call `libusb_handle_events` it takes this lock, and releases it when it returns. So even if you have multiple threads handling events (e.g. jumping in to make synchronous calls), the overall ordering of completions should be the same.

Yes, thank you @martinling . I am aware of this, thanks. I wish we could talk directly, it would make my point much easier to explain. Actually, now that the orderering is asserted, I am very well aware that the reminder of my issue in not really libusb-1.0 business, but let me explain:
Say that, in my implementation, in order to minimize the time spent in the callback function, all what the callback does is putting (could be copy or not) the received bytes into a queue. This queue will be processed by another thread, such that the libusb event handler thread can run as fast as possible.

Now: it there is only one single event handler thread, and since callbacks are made such as to maintain the order in which data is received on the bus, then the same order will be preserved in the queue mentionned before.

But, if there is 2 (or more) threads at some instant, for instance the regular dedicated event handler + another one caused by a "sync" call: one of those thread (say A) will receive the fisrt of the next available data chunk, and then sligtly later, another one  (say B) of those will receive the second next available data chunk. 

But from now on, nothing will gurantee that thread A will put its received bytes in the queue before thread B. One need to use an additional mechanism (such as numbers is User Data) in order to reorder chunks.

I decided to avoid this by rewriting the libusb built-in control transfer function with async transfer in a way that only my dedicated event handler thread process their callback. Its one way of handling the issue, there are other of course.

**Comment by martinling** [comment]
@sonatique I think that in the scenario you describe, the queue ordering is still guaranteed.

Your event handler thread and the various threads making sync calls cannot all hold the lock at the same time. Only one of them can actually be holding the lock and handling events. And your callbacks are called as part of the event handling.

Take a random example - say your dedicated thread (call it D) handles the completion of transfer 1. Then thread A comes in and takes over event handling and handles the completion of transfer 2. Meanwhile thread B makes another sync call, but A is already handling events and ends up handling the completion of transfer 3. Then B takes over and handles transfer 4, but then handling reverts back to D again for transfer 5.

Your callback ends up being called from thread D, then thread A twice, then B, then D again. But all those calls remain strictly sequenced because of the lock. The callback is called for transfer 1, then 2, then 3, then 4, then 5. And each call has to return before the next call can be made. The calls jump around between threads but they're not concurrent. If your callback puts the transfer data in a queue, then the queue will contain data from transfers 1-5 in order, regardless of what happens with the threads.

**Comment by sonatique** [comment]
@martinling Thanks a lot for you reply and sorry for my delayed answer.

Given current implemention, I am not sure that we can say that "only one [thread] can actually be holding the lock and handling events". I might be wrong, but looking at the code of "usbi_handle_transfer_completion" in io.c, I only see lock being held around the state change (setting state_flags).

I think that there is no lock around the actual transfer->callback (in which I would "add to queue" my things).
From there, I don't think we can say that all those calls in you example will remain strictly sequenced.
I see nothing preventing thread B to execute its "transfer->callback(...)" while thread A is still inside its own. 
So at one given instant I think I may have 2 (or more) threads concurrently executing my "add to queue" method. Even if I add synchronization mechanism to ensure my queue operations intergrity, there is no way to ensure that thread A will be able to "add to queue" before thread B, event if it started to execute these lines of code before.

I guess there is no lock around transfer->callback on purpose: if one callback function would take very long (or get dead-locked), this would prevent all other pending callback to be executed. Since the API (in particular the standard sync API) permit to submit and wait concurrently and independantly for several calls, I guess having a lock around the callback call would just not work in the general case.

As it is now, you may have one sync call blocked or taking very long while other are still happily executed on other threads.

For this reason, submitting several concurrent read transfers does not play nice (in the general case) with having multiple threads allowed to handle events. Especially, I figured out that calling the sync API from a random thread while having a dedicated thread handling all callbacks of my concurrently submitted transfers just screw everything, because the random thread could handle completion of my transfer concurrently with the dedicated thread, messing with the order, then go back to handling the result of the sync call. Another issue is the dedicated thread handling the result of the sync call, spending a long time doing it instead of taking care of my tranfers.
That's why I had to rewrite all my sync call for using only the dedicated thread. 

What do you think, am I completly wrong? Thanks!

**Comment by martinling** [comment]
@sonatique It sounds like you may be right, which would be bad news as it means a lot of code out there is subtly wrong!

If so, I wonder if there's some better way to fix this. It's painful to have to use a custom implementation of synchronous transfers just to enforce which thread does your event handling. And using `user_data` to reorder transfers is not great either, because it means you have to allocate extra space, and copy data around, to implement reordering.

@tormodvolden any thoughts?

Do we perhaps need some new API that lets you do:
```c
// In dedicated handler thread - bind the event handling to stay with this thread.
libusb_bind_event_handling(ctx);
while (!stop) {
    // All event handling for this device will be done here, so no transfer callbacks can run concurrently.
    libusb_handle_events(ctx);
}
libusb_unbind_event_handling(ctx);
```
...with the synchronous API functions modified such that they will play along, and not try to take over handling themselves?

**Comment by mcuee** [comment]
@tormodvolden I have added the `core` label and keep the `Documentation` label. This is one area I feel libusb needs better documentation and better examples. I have never been able to understand this well myself. Please share your thoughts about this issue and the proposal from Martin.

**Comment by benzea** [comment]
Hmm, I wonder why a reordering is occuring though. The events should be firing in the correct order (libusb internal locking), it is just random which thread ends up calling the callback. If reordering occurs, then I would expect that something else is going wrong in the application logic (which is completely plausible as e.g. an event on a separate thread might be queued while it is immediately processed if it is on the correct thread).

> It's painful to have to use a custom implementation of synchronous transfers just to enforce which thread does your event handling.
> ...
> Do we perhaps need some new API that lets you do: ... libusb_bind_event_handling(ctx); ...

I think you can already do that by using `libusb_try_lock_events` together with `libusb_handle_events_locked` (or the related functions for full event loop integration).

**Comment by martinling** [comment]
> Hmm, I wonder why a reordering is occuring though. The events should be firing in the correct order (libusb internal locking), it is just random which thread ends up calling the callback.

The theory that @sonatique put forward was that re-ordering and concurrency of callbacks can occur because the callbacks are not called with the lock held.

However, I'm looking at the code now and I think that @sonatique has misunderstood the locking. They mentioned `usbi_handle_transfer_completion` and saw locking only around the flag change in that function. But that is only the individual transfer lock, not the event waiters lock. The event locking occurs outside that function, further up the call stack.

The `usbi_handle_transfer_completion` function is called from the backend's `.handle_events` hook, which is called from the common `handle_events` helper in `io.c`. That `handle_events` helper is documented to assume that no other thread is concurrently running it. And it is called in only one place: inside `libusb_handle_events_timeout_completed`, which calls it with the event waiters lock held. All the other variants of `libusb_handle_events_*` are implemented as wrappers around `libusb_handle_events_timeout_completed`.

So I think what I said earlier in [this comment](https://github.com/libusb/libusb/issues/1077#issuecomment-1246078645) is correct, and that two callbacks can never run concurrently, or out of order, despite the fact that the event handling may jump around between threads.

Yet @sonatique was apparently seeing something like this happen, and needed to fix their code to prevent it. It would be good to know more about that, but it may just be a bug on the application side.

I think it might still be useful to have a way to force event handling to stay in a specific thread, though - e.g. if a callback needs to access thread-local variables, thread-owned resources, or to stay pinned to a core for performance reasons.

> I think you can already do that by using `libusb_try_lock_events` together with `libusb_handle_events_locked` (or the related functions for full event loop integration).

Yes, but that means you need to implement the polling part of the event loop yourself, and that's currently only possible on operating systems that use file descriptors. On Windows, event sources are represented as a `HANDLE`, not a file descriptor, and there's no `poll` - so `libusb_get_pollfds` just returns NULL. There's no equivalent Windows-specific function to get event handles for use with `WaitForMultipleObjects`, which is the equivalent API there.

There was a proposal [here](https://github.com/libusb/libusb/issues/252#issuecomment-299331116) in #252 to add a cross-platform API to access other types of event sources, but it hasn't happened as yet. Most projects have just decided to run a separate handler thread rather than trying to integrate libusb into their own event loop. Sigrok hung onto the latter approach for many years, but eventually gave up due to this limitation, and runs a handler thread in the background on Windows only.

**Comment by martinling** [comment]
Oh, sorry, I'm wrong about that last bit - you can indeed use `libusb_handle_events_locked` to force handling to a specific thread, without doing your own polling, by passing it a non-zero timeout.

**Comment by sonatique** [comment]
Thank you everybody! I think we had a very interesting discussion here, at least from my point of view because I learned a lot. Thank you all for your time and efforts in investigating and replying. 

I will reconsider all my code with the new light that this discussion shined on this matter, and try to find the real reason behind my issues, since it is quite clear now that they cannot come from libusb itself.

I still think that a good working reference example showing how to properly use multiple submitted bulk "in" transfers would help newcomers and make a statment about the fact that this is really a supported practice.

Thanks again. Of course I will come back here if I happen to find anything new relevant to the subjet.

**Comment by martinling** [comment]
Yes, queuing up multiple transfers definitely needs some explanatory documentation. This was also raised recently in #1205.

And as this discussion shows, what guarantees exist are far from clear, even to those working on the project.

Although we've confirmed that the event waiters lock prevents concurrent callbacks, it's still not 100% clear whether the order in which completions are handled is guaranteed to be the order in which they were submitted and executed on the bus. I think that for transfers submitted on the same endpoint the ordering ought to be maintained, but I'd like to be certain about whether that's guaranteed across all backends and situations.

**Comment by lujnan** [comment]
The root cause is that in different threads, synchronous and asynchronous APIs are mixed
In the composite device, even if the dup(fd) method is used, there is no way to prevent different threads from returning to reap_for_handle, and the thread will be scheduled by the system at this point in time, which leads to sequence disorder

**Comment by ecejeff** [comment]
Is the conclusion here that when multiple transfers are submitted to a single endpoint, libusb will not guarantee that the callbacks happen in order?

Is there any chance of libusb adding that guarantee? When callbacks can only happen in order it dramatically simplifies the user code.

**Comment by martinling** [comment]
It seems that nobody has been willing to authoritatively say that this guarantee is present yet, so let me try to do that.

I believe that if you submit multiple libusb transfers on the same endpoint from the same process, it is guaranteed that the execution order of completion callbacks matches the order of submission, because:

1. The ordering of completion notifications from the OS is guaranteed by the USB standard.
2. The ordering of processing those notifications in libusb is guaranteed by the event waiters lock.

The evidence for (1) will take some explanation because this guarantee is never explicitly stated in one single place in the standard. However, the standard makes other guarantees which, taken together, mean that no other interpretation can be possible. I will refer to the USB 2.0 standard here.

A "transfer" in libusb corresponds to an "I/O Request Packet" (IRP) in the standard - this is the generic name for the OS level transfer in the USB stack, e.g. a URB on Linux. IRPs are associated with "Pipes" (section 5.3.2). For IN and OUT endpoints, these are "Stream Pipes" (5.3.2.1). For the control endpoint, this is a "Message Pipe" (5.3.2.2).

The descriptions of these make it clear that requests are strictly ordered. For stream pipes :

> Stream pipes deliver data in the data packet portion of bus transactions with no USB-required structure on the data content. **Data flows in at one end of a stream pipe and out the other end in the same order**. Stream pipes are always uni-directional in their communication flow.
>
> Data flowing through a stream pipe is expected to interact with what the USB believes is a single client. The USB System Software is not required to provide synchronization between multiple clients that may be using the same stream pipe. **Data presented to a stream pipe is moved through the pipe in sequential order: first-in, first-out**.

And for message pipes:

> The USB System Software ensures that multiple requests are not sent to a message pipe concurrently. A device is required to service only a single message request at a time per message pipe. **Multiple software clients on the host can make requests via the Default Control Pipe, but they are sent to the device in a first- in, first-out order.** A device can control the flow of information during the Data and Status stages based on its ability to respond to the host transactions (refer to Chapter 8 for more details).

IRPs are reported as complete when their underlying bus transactions complete (5.3.2):

> A software client normally requests data transfers via I/O Request Packets (IRPs) to a pipe and then either waits or is notified when they are completed. Details about IRPs are defined in an operating system-specific manner. This specification uses the term to simply refer to an identifiable request by a software client to move data between itself (on the host) and an endpoint of a device in an appropriate direction. A software client can cause a pipe to return all outstanding IRPs if it desires. **The software client is notified that an IRP has completed when the bus transactions associated with it have completed either successfully or due to errors.**

The crucial passage when it comes to ordering is this one: when an IRP fails due to a bus error, _all subsequent IRPs are retired_, and the client must recover from the error before proceeding:

> If a non-isochronous pipe encounters a condition that causes it to send a STALL to the host (refer to Chapter 8) or three bus errors are encountered on any packet of an IRP, the IRP is aborted/retired, **all outstanding IRPs are also retired, and no further IRPs are accepted until the software client recovers from the condition** (in an implementation-dependent way) and acknowledges the halt or error condition via a USBD call.

My argument is basically that this approach to error handling cannot possibly work if IRP completion notifications may be received out-of-order, and therefore, they must be guaranteed to be received in-order.

This is consistent with observed behaviour on all operating systems over many years by projects relying on this assumption, including libhackrf, libsigrok/fx2lafw, etc.

What remains to be shown is (2): that libusb will always process these completions and execute their callbacks in the same order that the notifications arrive from the OS. The argument for that is simple: we do it only in one place, and doing so requires holding a lock.

From the [_Event handling_](https://libusb.sourceforge.io/api-1.0/group__libusb__asyncio.html#asyncevent) section of the docs:
> All event handling is performed by whichever thread calls the [libusb_handle_events()](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327) function. libusb does not invoke any callbacks outside of this context. Consequently, any callbacks will be run on the thread that calls the [libusb_handle_events()](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327) function.

The rest of the nitty-gritty about the event waiters lock is addressed in the [Multi-threaded applications and asynchronous I/O](https://libusb.sourceforge.io/api-1.0/libusb_mtasync.html) section.

To summarise: the guarantee is present because: the standard _requires_ the ordering, so the OS _provides_ the ordering, and libusb _maintains_ the ordering.

**Comment by ecejeff** [comment]
Thanks for taking the time to respond in such a detailed way. I really appreciate it! 

I 100% agree with your interpretation that the USB spec mandates it, and any OS will retain the ordering.

> This is consistent with observed behaviour on all operating systems over many years by projects relying on this assumption, including libhackrf, libsigrok/fx2lafw, etc.

Libusb makes this the "obvious" way to write transfer callbacks, but never states that it will work. Unfortunately, libusb also makes other obvious things illegal (at least in practice) like mixing synchronous and asynchronous calls in the same program as noted in #1273. So it being obvious doesn't mean it's correct.

My old code had to deal with #244 and #126 (both fixed now), so this simple method of handling callback data hasn't been available to me until my current rewrite. I'm realizing that I don't have to deal with spurious cancels but I might still have to deal with reordering. Looking into that lead me to this bug report talking about exactly that issue.

> What remains to be shown is (2): that libusb will always process these completions and execute their callbacks in the same order that the notifications arrive from the OS. The argument for that is simple: we do it only in one place, and doing so requires holding a lock.
> 
> From the [_Event handling_](https://libusb.sourceforge.io/api-1.0/group__libusb__asyncio.html#asyncevent) section of the docs:
> 
> > All event handling is performed by whichever thread calls the [libusb_handle_events()](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327) function. libusb does not invoke any callbacks outside of this context. Consequently, any callbacks will be run on the thread that calls the [libusb_handle_events()](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327) function.
> 
> The rest of the nitty-gritty about the event waiters lock is addressed in the [Multi-threaded applications and asynchronous I/O](https://libusb.sourceforge.io/api-1.0/libusb_mtasync.html) section.

It seems to me that all this promises is that the callbacks will execute in a sequence without overlapping (which is good). It doesn't promise that the callbacks will execute in the same sequence that the OS delivers results. 

> To summarise: the guarantee is present because: the standard _requires_ the ordering, 

Agree. Your citations are pretty explicit about this.

> so the OS _provides_ the ordering, 

Agree, though I don't have proof, it just makes a lot of sense.

> and libusb _maintains_ the ordering.

Not clear to me yet, and not in any doc I've found.

Looking at the source makes me think the `completed_transfers` list will be delivered in order, but it's hard to know if each backend queues to this in-order. I think it's true on Windows, Linux and Mac, but I'm not familiar enough with the OS specific details to say with certainty.

**Comment by martinling** [comment]
> Libusb makes this the "obvious" way to write transfer callbacks, but never states that it will work. Unfortunately, libusb also makes other obvious things illegal (at least in practice) like mixing synchronous and asynchronous calls in the same program as noted in [#1273](https://github.com/libusb/libusb/issues/1273). So it being obvious doesn't mean it's correct.

I don't fully follow the discussion in #1273, but I do note that it involves transfers on two different endpoints (the control endpoint, and an interrupt endpoint). The USB standard doesn't provide any guarantees about ordering across different endpoints, only between data and IRPs on the same endpoint, so I don't think libusb can provide any guarantee either. That's out of the scope of the discussion above, though.

> Looking at the source makes me think the completed_transfers list will be delivered in order, but it's hard to know if each backend queues to this in-order. I think it's true on Windows, Linux and Mac, but I'm not familiar enough with the OS specific details to say with certainty.

If completions are received in order from the OS, and a thread must hold the event waiters lock to receive them from the OS - such that only one thread can be doing so at once - then how would they ever become out of order in the first place?

I mean, one could concoct a backend that re-queues them in a random order, but so long as we don't do that, the usage of the event waiters lock provides a cross-platform architectural guarantee that we maintain whatever order is provided by the OS.

**Comment by diabolo38** [comment]
mix sync and asynchronous  api can work maybe limited to some cases and only if  using libusb_handle_events_completed   with proper completed flags
I have it working on all os (win, linux x86, linux arm64, mac x86)  it make simple and optimized host to device command  with response back exchange. The assync read to read ep  is queued  first then sync  wr to wr ep is performed. thus write to read latency is minimal without need to handle complex cancellation case or race /order in completion (but the event completed flags).

Also i think  even assynch completion arrive back in order on bus then os level  if  you'r multiple thread are waiting on same lock/mutex object  to somehow manage the  data then you can't quite assume the thread lock acquire order. Even without lock/mutex wake thread  completion call order   is unpredictable, it is  dependent on os ,scheduler and  cpu(s) load .

**Comment by martinling** [comment]
> Also i think even assynch completion arrive back in order on bus then os level if you'r multiple thread are waiting on same lock/mutex object to somehow manage the data then you can't quite assume the thread lock acquire order.

If you have multiple threads trying to run `libusb_handle_events` and waiting for the event waiters lock, the acquire order _doesn't matter_. The callback order will be correct, it's only _which threads_ the callbacks run in that will be random.

**Comment by ecejeff** [comment]
> If completions are received in order from the OS, and a thread must hold the event waiters lock to receive them from the OS - such that only one thread can be doing so at once - then how would they ever become out of order in the first place?

For example, a backend that selects/polls to see which transfers are complete then iterates an unsorted list of transfers (like a hash set for O(1) lookup) and queues each that has been marked as finished. Most of the time this would return things in order, but if multiple transfers finish during the same iteration (could happen under heavy CPU load) then the completion order would be essentially random.

I have no idea if any OS backend does this, but it's pretty easy for me to imagine. Moreover, with no guarantee about ordering, this is perfectly valid behavior.

I would like to see the project docs explicitly say which side they're coming down on. Either state there is no ordering and user code needs to account for this, or state that there is an ordering (for transfers to the same device endpoint) and user code can rely on it. 

I'm willing to make the PR to update the docs either way (if anyone wants me to), but I don't actually know which way the project wants to go.

**Comment by martinling** [comment]
That's fair. Then somebody still needs to look at each backend and make a convincing argument that it won't ever reorder the completion notifications from the OS, for transfers on the same endpoint.

**Comment by ecejeff** [comment]
My (LLM assisted) search through the various backends shows that the SunOS backend definitely reorders transfers. The other OSs didn't yield immediate positive or negative results, and I gave up refining my search after finding the SunOS result.

SunOS:
Because the AIO is submitted with `SIGEV_THREAD`, the OS scheduler is free to call the various `sunos_async_callback` runs in any order, which means `usbi_signal_transfer_completion` can be called in any order for the transfers. QED I guess.

Any portable code using LibUSB today needs to assume transfer callbacks are called in random order.
